File: src/scte224_service/parser.py
"""XML parser that converts SCTE-224 media schedules into domain models."""

from __future__ import annotations

from pathlib import Path
import xml.etree.ElementTree as ET

from .models import (
    ApplyAction,
    MatchSignal,
    Media,
    MediaContent,
    MediaPoint,
    Policy,
    RemoveAction,
    SignalAssertion,
    ViewingPolicy,
)
from .utils import parse_bool, parse_datetime, parse_duration


NS = {
    "scte": "http://www.scte.org/schemas/224",
    "action": "http://qplive.com/schemas/224/action",
    "xlink": "http://www.w3.org/1999/xlink",
}


class MediaParser:
    """Parses SCTE-224 XML or JSON payloads into Python domain objects."""

    def parse_file(self, path: str | Path) -> Media:
        xml_text = Path(path).read_text(encoding="utf-8")
        return self.parse_xml(xml_text)

    def parse_xml(self, xml_text: str) -> Media:
        root = ET.fromstring(xml_text)
        if root.tag.split("}")[-1] != "Media":  # pragma: no cover - defensive
            raise ValueError("Root element must be <Media>")

        media = Media(
            id=root.get("id"),
            description=root.get("description"),
            last_updated=parse_datetime(root.get("lastUpdated")),
            effective=parse_datetime(root.get("effective")),
            expires=parse_datetime(root.get("expires")),
        )

        for mp_elem in root.findall("scte:MediaPoint", NS):
            media.media_points.append(self._parse_media_point(mp_elem))

        return media

    def _parse_media_point(self, element: ET.Element) -> MediaPoint:
        media_point = MediaPoint(
            id=element.get("id"),
            description=element.get("description"),
            effective=parse_datetime(element.get("effective")),
            expires=parse_datetime(element.get("expires")),
            match_time=parse_datetime(element.get("matchTime")),
            expected_duration=parse_duration(element.get("expectedDuration")),
            reusable=parse_bool(element.get("reusable")) if element.get("reusable") else False,
        )

        for alt in element.findall("scte:AltID", NS):
            alt_type = alt.get("type", "default")
            media_point.alt_ids[alt_type] = (alt.text or "").strip()

        for match_signal in element.findall("scte:MatchSignal", NS):
            media_point.match_signals.append(self._parse_match_signal(match_signal))

        for apply in element.findall("scte:Apply", NS):
            media_point.apply_actions.append(self._parse_apply(apply))

        for remove in element.findall("scte:Remove", NS):
            media_point.remove_actions.append(self._parse_remove(remove))

        return media_point

    def _parse_match_signal(self, element: ET.Element) -> MatchSignal:
        match_signal = MatchSignal(
            match=element.get("match", "ALL").upper(),
            schema=element.get("schema", ""),
        )

        for assertion in element.findall("scte:Assert", NS):
            text = (assertion.text or "").strip()
            if not text:
                continue
            match_signal.assertions.append(self._parse_assertion(text))

        return match_signal

    def _parse_assertion(self, expression: str) -> SignalAssertion:
        # Very small subset of XPath used in the sample schedule.
        expression = expression.strip()
        if expression.startswith("/"):
            expression = expression[1:]

        if "contains" in expression:
            path_part, contains_clause = expression.split("contains", 1)
            path = [p for p in path_part.split("/") if p]
            contains_clause = contains_clause.strip("()")
            attr_name, value = contains_clause.split(",")
            attr_name = attr_name.strip().lstrip("@")
            value = value.strip().strip("'\"")
            return SignalAssertion(path=path, contains={attr_name: value})

        if "[" not in expression:
            path = [p for p in expression.split("/") if p]
            return SignalAssertion(path=path)

        path_part, predicate_part = expression.split("[", 1)
        path = [p for p in path_part.split("/") if p]
        predicate_part = predicate_part.rstrip("]")

        equals = {}
        for clause in predicate_part.split(" and "):
            clause = clause.strip()
            if not clause:
                continue
            if "=" not in clause:
                continue
            name, value = clause.split("=", 1)
            name = name.strip().lstrip("@")
            value = value.strip().strip("'\"")
            equals[name] = value

        return SignalAssertion(path=path, equals=equals)

    def _parse_apply(self, element: ET.Element) -> ApplyAction:
        priority = int(element.get("priority", "0"))
        duration = parse_duration(element.get("duration"))
        policy_elem = element.find("scte:Policy", NS)
        if policy_elem is None:
            raise ValueError("<Apply> is missing nested <Policy>")

        policy = self._parse_policy(policy_elem)
        return ApplyAction(priority=priority, duration=duration, policy=policy)

    def _parse_remove(self, element: ET.Element) -> RemoveAction:
        policy_elem = element.find("scte:Policy", NS)
        if policy_elem is None:
            raise ValueError("<Remove> is missing nested <Policy>")
        href = policy_elem.get(f"{{{NS['xlink']}}}href")
        if not href:
            raise ValueError("<Remove><Policy> must provide xlink:href")
        return RemoveAction(policy_href=href)

    def _parse_policy(self, element: ET.Element) -> Policy:
        viewing_policies = []
        for vp_elem in element.findall("scte:ViewingPolicy", NS):
            viewing_policies.append(self._parse_viewing_policy(vp_elem))

        return Policy(
            id=element.get("id", ""),
            viewing_policies=viewing_policies,
        )

    def _parse_viewing_policy(self, element: ET.Element) -> ViewingPolicy:
        audience_elem = element.find("scte:Audience", NS)
        audience_id = audience_elem.get("id") if audience_elem is not None else None

        content_elem = element.find("action:Content", NS)
        if content_elem is None:
            content = MediaContent()
        else:
            content = MediaContent(
                source_uri=self._find_text(content_elem, "action:SourceURI"),
                alternate_uri=self._find_text(content_elem, "action:AlternateURI"),
                fallback_uri=self._find_text(content_elem, "action:FallbackURI"),
            )

        return ViewingPolicy(
            id=element.get("id", ""),
            audience_id=audience_id,
            content=content,
        )

    @staticmethod
    def _find_text(parent: ET.Element, path: str) -> str | None:
        node = parent.find(path, NS)
        if node is None or node.text is None:
            return None
        return node.text.strip()

File: internal/parser/parser.go
package parser

import (
	"encoding/xml"
	"errors"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/example/scte224service/internal/domain"
)

// MediaParser reads SCTE-224 Media documents and builds domain models.
type MediaParser struct{}

// ParseFile reads an XML document from disk.
func (p *MediaParser) ParseFile(path string) (*domain.Media, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open media file: %w", err)
	}
	defer file.Close()
	return p.Parse(file)
}

// Parse consumes an io.Reader containing SCTE-224 XML.
func (p *MediaParser) Parse(r io.Reader) (*domain.Media, error) {
	var doc mediaDocument
	decoder := xml.NewDecoder(r)
	decoder.CharsetReader = charsetReader
	if err := decoder.Decode(&doc); err != nil {
		return nil, fmt.Errorf("decode media document: %w", err)
	}
	return translateMedia(doc)
}

// charsetReader handles UTF-8 only but keeps extension point.
func charsetReader(_ string, input io.Reader) (io.Reader, error) {
	return input, nil
}

type mediaDocument struct {
	XMLName     xml.Name     `xml:"{http://www.scte.org/schemas/224}Media"`
	ID          string       `xml:"id,attr"`
	Description string       `xml:"description,attr"`
	LastUpdated string       `xml:"lastUpdated,attr"`
	Effective   string       `xml:"effective,attr"`
	Expires     string       `xml:"expires,attr"`
	MediaPoints []mediaPoint `xml:"MediaPoint"`
}

type mediaPoint struct {
	ID               string          `xml:"id,attr"`
	Description      string          `xml:"description,attr"`
	Effective        string          `xml:"effective,attr"`
	Expires          string          `xml:"expires,attr"`
	MatchTime        string          `xml:"matchTime,attr"`
	ExpectedDuration string          `xml:"expectedDuration,attr"`
	Reusable         string          `xml:"reusable,attr"`
	AltIDs           []altID         `xml:"AltID"`
	MatchSignals     []matchSignal   `xml:"MatchSignal"`
	Apply            []applyElement  `xml:"Apply"`
	Remove           []removeElement `xml:"Remove"`
}

type altID struct {
	Type  string `xml:"type,attr"`
	Value string `xml:",chardata"`
}

type matchSignal struct {
	Match   string         `xml:"match,attr"`
	Schema  string         `xml:"schema,attr"`
	Asserts []signalAssert `xml:"Assert"`
}

type signalAssert struct {
	Expression string `xml:",chardata"`
}

type applyElement struct {
	Priority string        `xml:"priority,attr"`
	Duration string        `xml:"duration,attr"`
	Policy   policyElement `xml:"Policy"`
}

type removeElement struct {
	Policies []removePolicy `xml:"Policy"`
}

type removePolicy struct {
	Href string `xml:"{http://www.w3.org/1999/xlink}href,attr"`
}

type policyElement struct {
	ID              string          `xml:"id,attr"`
	ViewingPolicies []viewingPolicy `xml:"ViewingPolicy"`
}

type viewingPolicy struct {
	ID       string          `xml:"id,attr"`
	Audience audienceElement `xml:"Audience"`
	Content  contentElement  `xml:"{http://qplive.com/schemas/224/action}Content"`
}

type audienceElement struct {
	ID string `xml:"id,attr"`
}

type contentElement struct {
	SourceURI    string `xml:"{http://qplive.com/schemas/224/action}SourceURI"`
	AlternateURI string `xml:"{http://qplive.com/schemas/224/action}AlternateURI"`
	FallbackURI  string `xml:"{http://qplive.com/schemas/224/action}FallbackURI"`
}

func translateMedia(doc mediaDocument) (*domain.Media, error) {
	if doc.ID == "" {
		return nil, errors.New("media id is required")
	}
	media := &domain.Media{
		ID:          doc.ID,
		Description: doc.Description,
	}
	var err error
	if media.LastUpdated, err = parseTime(doc.LastUpdated); err != nil {
		return nil, fmt.Errorf("media lastUpdated: %w", err)
	}
	if media.Effective, err = parseTime(doc.Effective); err != nil {
		return nil, fmt.Errorf("media effective: %w", err)
	}
	if media.Expires, err = parseTime(doc.Expires); err != nil {
		return nil, fmt.Errorf("media expires: %w", err)
	}
	for _, mp := range doc.MediaPoints {
		converted, err := translateMediaPoint(mp)
		if err != nil {
			return nil, fmt.Errorf("media point %s: %w", mp.ID, err)
		}
		media.MediaPoints = append(media.MediaPoints, converted)
	}
	return media, nil
}

func translateMediaPoint(mp mediaPoint) (*domain.MediaPoint, error) {
	if mp.ID == "" {
		return nil, errors.New("id attribute is required")
	}
	result := &domain.MediaPoint{
		ID:          mp.ID,
		Description: mp.Description,
		AltIDs:      make(map[string]string),
		Reusable:    strings.EqualFold(mp.Reusable, "true"),
	}
	var err error
	if result.Effective, err = parseTime(mp.Effective); err != nil && mp.Effective != "" {
		return nil, fmt.Errorf("effective: %w", err)
	}
	if result.Expires, err = parseTime(mp.Expires); err != nil && mp.Expires != "" {
		return nil, fmt.Errorf("expires: %w", err)
	}
	if result.MatchTime, err = parseTime(mp.MatchTime); err != nil && mp.MatchTime != "" {
		return nil, fmt.Errorf("matchTime: %w", err)
	}
	if result.ExpectedDuration, err = parseDuration(mp.ExpectedDuration); err != nil && mp.ExpectedDuration != "" {
		return nil, fmt.Errorf("expectedDuration: %w", err)
	}
	for _, alt := range mp.AltIDs {
		if alt.Type != "" {
			result.AltIDs[alt.Type] = strings.TrimSpace(alt.Value)
		}
	}
	for _, ms := range mp.MatchSignals {
		result.MatchSignals = append(result.MatchSignals, translateMatchSignal(ms))
	}
	for _, apply := range mp.Apply {
		action, err := translateApply(apply)
		if err != nil {
			return nil, err
		}
		result.ApplyActions = append(result.ApplyActions, action)
	}
	for _, rem := range mp.Remove {
		for _, pol := range rem.Policies {
			result.RemoveActions = append(result.RemoveActions, &domain.RemoveAction{PolicyRef: pol.Href})
		}
	}
	return result, nil
}

func translateMatchSignal(ms matchSignal) *domain.MatchSignal {
	result := &domain.MatchSignal{Match: ms.Match, Schema: ms.Schema}
	for _, assertion := range ms.Asserts {
		result.Asserts = append(result.Asserts, &domain.SignalAssertion{Expression: strings.TrimSpace(assertion.Expression)})
	}
	return result
}

func translateApply(apply applyElement) (*domain.ApplyAction, error) {
	priority, err := parseInt(apply.Priority, 0)
	if err != nil {
		return nil, fmt.Errorf("apply priority: %w", err)
	}
	duration, err := parseDuration(apply.Duration)
	if err != nil && apply.Duration != "" {
		return nil, fmt.Errorf("apply duration: %w", err)
	}
	policy := &domain.Policy{ID: apply.Policy.ID}
	for _, vp := range apply.Policy.ViewingPolicies {
		policy.ViewingPolicies = append(policy.ViewingPolicies, translateViewingPolicy(vp))
	}
	return &domain.ApplyAction{
		Priority: priority,
		Duration: duration,
		Policy:   policy,
	}, nil
}

func translateViewingPolicy(vp viewingPolicy) *domain.ViewingPolicy {
	content := &domain.Content{
		SourceURI:    strings.TrimSpace(vp.Content.SourceURI),
		AlternateURI: strings.TrimSpace(vp.Content.AlternateURI),
		FallbackURI:  strings.TrimSpace(vp.Content.FallbackURI),
	}
	audienceID := strings.TrimSpace(vp.Audience.ID)
	return &domain.ViewingPolicy{
		ID:         vp.ID,
		AudienceID: audienceID,
		Content:    content,
	}
}

func parseTime(value string) (time.Time, error) {
	if strings.TrimSpace(value) == "" {
		return time.Time{}, nil
	}
	if strings.HasSuffix(value, "Z") {
		value = strings.TrimSuffix(value, "Z") + "+00:00"
	}
	t, err := time.Parse(time.RFC3339, value)
	if err != nil {
		return time.Time{}, err
	}
	return t.UTC(), nil
}

func parseDuration(value string) (time.Duration, error) {
	if strings.TrimSpace(value) == "" {
		return 0, nil
	}
	if !strings.HasPrefix(value, "P") {
		return 0, fmt.Errorf("duration must start with 'P'")
	}
	value = strings.TrimPrefix(value, "P")
	var (
		total  time.Duration
		num    strings.Builder
		inTime bool
	)
	flush := func(multiplier time.Duration) {
		if num.Len() == 0 {
			return
		}
		v, _ := parseInt(num.String(), 0)
		total += time.Duration(v) * multiplier
		num.Reset()
	}
	for _, r := range value {
		switch r {
		case 'T':
			inTime = true
		case 'D':
			flush(24 * time.Hour)
		case 'H':
			if !inTime {
				return 0, fmt.Errorf("hours specified before time designator")
			}
			flush(time.Hour)
		case 'M':
			if !inTime {
				return 0, fmt.Errorf("month parsing not supported")
			}
			flush(time.Minute)
		case 'S':
			if !inTime {
				return 0, fmt.Errorf("seconds specified before time designator")
			}
			flush(time.Second)
		default:
			num.WriteRune(r)
		}
	}
	return total, nil
}

func parseInt(value string, fallback int) (int, error) {
	value = strings.TrimSpace(value)
	if value == "" {
		return fallback, nil
	}
	i, err := strconv.Atoi(value)
	if err != nil {
		return 0, err
	}
	return i, nil
}
